File: src/ecs/entity.lua
local Entity = {}

function Entity:new(id, manager, meta)
  local entity = {
    id = id,
    manager = manager,
    components = 0,
    systems = 0,
    meta = meta or {}
  }

  function entity:destroy()
    self.manager:removeEntity(self)
  end

  function entity:has(cmp)
    return bit.band(self.components, cmp._meta.flag) == cmp._meta.flag
  end

  function entity:as(cmp)
    return self.manager:getComponent(self, cmp)
  end

  function entity:addComponent(cmp)
    return self.manager:addComponent(self, cmp)
  end

  function entity:removeComponent(cmp)
    return self.manager:removeComponent(self, cmp)
  end

  function entity:setFlag(cmp)
    if self:has(cmp) then
      print('entity:' .. self.id .. ' already has ' .. cmp._meta.id)
      return
    end

    self.components = bit.bxor(self.components, cmp._meta.flag)
  end

  function entity:clearFlag(cmp)
    if not self:has(cmp) then
      print('entity:' .. self.id .. ' does not have ' .. cmp._meta.id)
      return
    end

    self.components = bit.band(self.components, bit.bnot(cmp._meta.flag))
  end

  return entity
end

return Entity

File: client2/src/ecs/Entity.ts
import { Component } from './component'
import { Aliasable } from './types'

type Manager = any

interface UserData {
  [k: string]: any
  blueprint?: string
}

export class Entity {
  components: number
  id: number
  manager: Manager
  systems: number
  userData: UserData

  constructor(id: number, manager: Manager, userData: UserData = {}) {
    this.components = 0
    this.id = id
    this.manager = manager
    this.systems = 0
    this.userData = userData
  }

  add = (component: Component) => {
    this.manager.addComponent(this, component)
  }

  addAll = (components: Component[]) => {
    components.forEach(component => this.add(component))
  }

  as = <T>(Class: Aliasable<T>) => {
    const cmp = this.manager.getComponent(this, Class)

    return cmp && ((cmp as unknown) as T)
  }

  clearFlag = (component: Component) => {
    if (!this.has(component)) {
      print(`entity ${this.id} does not have component ${component._flag}`)
    }

    this.components = bit.band(this.components, bit.bnot(component._flag))
  }

  destroy = () => {
    this.manager.removeEntity(this)
  }

  has = (component: Component) => {
    return bit.band(this.components, component._flag) === component._flag
  }

  remove = (component: Component) => {
    this.manager.removeComponent(this, component)
  }

  setFlag = (component: Component) => {
    if (this.has(component)) {
      print(`entity ${this.id} already has component ${this.id}`)
      return
    }

    this.components = bit.bxor(this.components, component._flag)
  }
}
